### PR TITLE
Build: Replace oldest-supported-numpy

### LINKED ISSUE #14915: Drop oldest-supported-numpy from build requirements when numpy 1.25 is released
@saimn said, "once 1.25 is out we can put it as minimum version in our build requirements," and drop `oldest-supported-numpy`. `@seberg` has also confirmed separately that this should work. cc @astropy/astropy-project-release-team 

https://numpy.org/devdocs/release/1.25.0-notes.html#by-default-the-exported-numpy-c-api-is-now-compatible-with-numpy-1-19 (https://github.com/numpy/numpy/pull/23528 and https://github.com/numpy/numpy/pull/23872)

https://numpy.org/devdocs/dev/depending_on_numpy.html#for-downstream-package-authors

https://github.com/astropy/astropy/blob/0e4ab57499b49f18f60b87baf5a74d964d6decde/pyproject.toml#L5

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to replace oldest-supported-numpy with numpy 1.25 that guarantee backward compatible C-API.

Fixes #14915

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 requires = ["setuptools",
             "setuptools_scm>=6.2",
             "cython==0.29.34",
-            "oldest-supported-numpy",
+            "numpy==1.25",
```
Comment (astrofrog) [inline pyproject.toml:5]: Maybe this should be >=1.25

Comment (astrofrog) [reply pyproject.toml:5]: I wonder if this should perhaps also include <2 though I haven't had time to check if the backward compatibility is 'forever' or could break in 2.0

Comment (astrofrog) [reply pyproject.toml:5]: "A major release is highly unusual (NumPy is still at version 1.xx) and if it happens it will likely indicate an ABI break"

I think we should perhaps do >=1.25,<2

Comment (pllim) [reply pyproject.toml:5]: I thought someone once said that we should always do `==` pins on the build requirements? Isn't that why we do `==` for cython? Then why cython but not numpy?

Comment (saimn) [reply pyproject.toml:5]: Agreed about the `>=1.25,<2.0` pin for Numpy, 
For Cython, Numpy uses `Cython>=0.29.34,<3.0`  which seems better to me than pinning a specific version.

Comment (pllim) [reply pyproject.toml:5]: But how can we know numpy 1.26 won't drop things what we need?

Comment (saimn) [reply pyproject.toml:5]: Yes, I think we can trust Numpy on this and go with `>=1.25,<2.0`.

**Comment by saimn** [comment]
Ideally we should also use
```
#define NPY_TARGET_VERSION NPY_1_21_API_VERSION
```
but I don't think we have an easy way to do that right now with extension-helpers ...

**Comment by pllim** [comment]
@saimn , do you want to take over? You probably knows exactly what to do in the loop and stuff. Thanks!

File: pyproject.toml (hunk 1 above)
Comment (pllim) [reply pyproject.toml]: OK I updated the pins but I am not so sure about touching `get_extensions`, so I'll leave that to you.

**Comment by saimn** [comment]
Well there are various things for which a `setup.py` is still necessary, cython coverage, profiling, debugging...

**Comment by astrofrog** [comment]
Can't we just add an option in extension helpers to do this automatically for all extensions?

**Comment by saimn** [comment]
@astrofrog - Yes, I guess this could be done.

File: setup.py [hunk 2]
```diff
@@ -66,4 +66,11 @@
 
 from extension_helpers import get_extensions  # noqa: E402
 
-setup(ext_modules=get_extensions())
+ext_modules = get_extensions()
+
+# Specify the minimum version for the Numpy C-API
+for ext in ext_modules:
+    if ext.include_dirs and "numpy" in ext.include_dirs[0]:
+        ext.define_macros.append(("NPY_TARGET_VERSION", "NPY_1_21_API_VERSION"))
```
Comment (pllim) [inline setup.py:74]: Does it crash if the extension code already has its own `NPY_TARGET_VERSION` somehow?

Comment (pllim) [reply setup.py:74]: Also... so much for getting rid of `setup.py` lol

File: pyproject.toml (hunk 1 above)
Comment (seberg) [reply pyproject.toml]: NumPy 2.0 breaks forward compatibility but provides backwards compatibility: A wheel compiled with 1.X cannot run with 2.X.  So you will have to compile wheels against 2.0 when it comes out.

I may introduce a `numpy2_compat` as a runtime dependency to run against 1.X, but I haven't made up my mind yet: it might be too much hassle for little payoff (also considering dev cycles).

I will note that the only real reason to add:
```
#define NPY_TARGET_VERSION NPY_1_21_API_VERSION
```
at this point is if you would actually want to use newer C-API, which you don't.

Comment (pllim) [reply pyproject.toml]: > A wheel compiled with 1.X cannot run with 2.X.

That is bad news for downstream packages that pull dev wheels for both `numpy` and `astropy`. How do we generate a dev wheel of astropy that is compatible with dev wheel of numpy before numpy 2.0 comes out?

Comment (seberg) [reply pyproject.toml]: Hmmmpf, I guess you will have to compile it against our dev wheel, you don't even have to pass tests on it to do that.

The problem is, only if you use the 2.x wheel, can I provide updated headers to ensure compatibility (and force you to remove API that is I want to stop dragging around).

Comment (seberg) [reply pyproject.toml]: Sorry, somehow slipped me:  Right now, NumPy doesn't actually enforce this!  For precisely that reason.  As long as we just remove things for the most part, I am happy to gamble that the wheels are effectively ABI compatible (even if that is a lie).

There might come a day where we want to do something that makes this impossible, but I am not sure that will come...
As soon as there is an actual NumPy `rc1`, that would be enforced though.

Comment (pllim) [reply pyproject.toml]: > you will have to compile it against our dev wheel

I guess it is theoretically possible for our unreleased 6.0.dev. But we cannot do that for release branches; so does this mean we have to pin numpy maxversion for v5.3.x ? Though I also have no seen any failures related to this in our CI set up. My head hurts. 😆

Comment (seberg) [reply pyproject.toml]: Sorry, yes, just repeating in case it got lost:  The NumPy dev wheels do *not* actually enforce this, they just pretend they are still 1.x ABI (it's a lie, but who cares).  Maybe that will just stay that way, maybe I will find a big enough reason to break that "promise".

**Comment by pllim** [comment]
So... Are we keeping the stuff in `setup.py` here or no? I got lost in the discussions.

**Comment by pllim** [comment]
I am revisiting Python 3.12 related stuff. I realized that some upstream stuff with oldest-supported-numpy is giving us troubles if we have to build them from source. So this means we're affected our downstream the same way if for whatever reason they cannot use our dev wheels (we don't have py312 wheels yet but I want to make it happen).

tl;dr -- Can I merge this???

**Review by astrofrog — APPROVED** [review]
Go forth and merge! I will think about how to avoid the setup.py code here, I think we can do it via extension-helpers. But this is fine for now.

**Comment by pllim** [comment]
Failures are unrelated but I do have to wait for #15197 to get in first and rebase, just in case.

**Comment by pllim** [comment]
Let's run all the exotic stuff too, just in case.

File: pyproject.toml [hunk 3]
```diff
@@ -2,7 +2,7 @@
 requires = ["setuptools",
             "setuptools_scm>=6.2",
             "cython==0.29.34",
-            "oldest-supported-numpy",
+            "numpy>=1.25,<2",
```
Comment (neutrinoceros) [inline pyproject.toml:5]: I would recommend adding a temporary version cap on the runtime dependency as well, see https://github.com/scipy/oldest-supported-numpy/issues/76#issuecomment-1628819676

Comment (neutrinoceros) [reply pyproject.toml:5]: Yes I mean `install_requires`, and it won't be blocking for future-compatibility tests because `2.0.0dev0` compares as `< 2.0` :)

Comment (pllim) [reply pyproject.toml:5]: I guess I do not understand the logic of allowing 2.0.dev but preventing users from using the actual 2.0 later.

Comment (pllim) [reply pyproject.toml:5]: Thanks for your concern, which is valid. But I would like to discuss it separate from this one. Can you please propose a PR at your convenience?

I don't mind max pinning numpy in release branches, but I would rather not do it in the main branch. I *think* scipy only pin maxversion when they branch for release (at least the last I look and I don't look very often).

Comment (neutrinoceros) [reply pyproject.toml:5]: done at #15206 !
